File: bindings/python/src/kortecx/motes.py
"""Batch B per-mote definition view — ``GetMoteDetail``. DISPLAY-ONLY (SN-8):
the capped definition summary the coordinator persisted at admission, resolved
by ``mote_def_hash``; nothing here authorizes anything. A mote that has not
committed (or was admitted by a pre-Batch-B binary) answers
``def_found=False`` honestly. Kept in its own module per the
module-per-concern discipline (the ``runs.py`` pattern).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional

from . import hexids
from .v1 import gateway_pb2 as _g

_ND_CLASS_NAMES = {1: "PURE", 2: "READ_ONLY_NONDET", 3: "WORLD_MUTATING"}
_EFFECT_PATTERN_NAMES = {
    1: "IdempotentByConstruction",
    2: "StageThenCommit",
    3: "ValidateThenCommit",
}


def nd_class_name(nd: int) -> str:
    """Display name for a wire ``NdClass`` discriminant (CLI-parity strings)."""
    return _ND_CLASS_NAMES.get(nd, "UNKNOWN")


def effect_pattern_name(ep: int) -> str:
    """Display name for a wire ``EffectPattern`` discriminant (CLI-parity strings)."""
    return _EFFECT_PATTERN_NAMES.get(ep, "UNKNOWN")


@dataclass(frozen=True)
class MoteConfigItem:
    """One capped config entry of a Mote definition (opaque display bytes)."""

    key: str
    value: bytes  # possibly truncated
    truncated: bool
    full_len: int

    @classmethod
    def from_proto(cls, e: "_g.MoteConfigEntry") -> "MoteConfigItem":
        return cls(key=e.key, value=e.value, truncated=e.truncated, full_len=e.full_len)

    def to_dict(self) -> dict:
        """The CLI ``--json`` parity shape."""
        return {
            "key": self.key,
            "value_hex": self.value.hex(),
            "truncated": self.truncated,
            "full_len": self.full_len,
        }


@dataclass(frozen=True)
class MoteDetail:
    """The capped, display-only definition summary of one Mote."""

    mote_id: str  # hex
    mote_def_hash: str  # hex; EMPTY until the Mote commits
    def_found: bool  # False: uncommitted, or admitted by a pre-Batch-B binary
    step_kind: str  # "pure"|"model"|"exec"|"shaper"|"critic"|"react-turn" (display)
    model_id: str
    prompt: str  # config_subset["prompt"], capped server-side
    prompt_truncated: bool
    config_subset: List[MoteConfigItem]
    tool_contract: Dict[str, str]  # tool name -> pinned version
    logic_ref: str  # hex
    nd_class: int
    effect_pattern: int
    critic_for: Optional[str]  # hex producer id, or None
    is_topology_shaper: bool
    schema_version: int

    @classmethod
    def from_proto(cls, d: "_g.MoteDetail") -> "MoteDetail":
        return cls(
            mote_id=hexids.encode(d.mote_id),
            mote_def_hash=hexids.encode(d.mote_def_hash),
            def_found=d.def_found,
            step_kind=d.step_kind,
            model_id=d.model_id,
            prompt=d.prompt,
            prompt_truncated=d.prompt_truncated,
            config_subset=[MoteConfigItem.from_proto(e) for e in d.config_subset],
            tool_contract=dict(d.tool_contract),
            logic_ref=hexids.encode(d.logic_ref),
            nd_class=d.nd_class,
            effect_pattern=d.effect_pattern,
            critic_for=hexids.encode(d.critic_for) if d.HasField("critic_for") else None,
            is_topology_shaper=d.is_topology_shaper,
            schema_version=d.schema_version,
        )

    @property
    def nd_class_name(self) -> str:
        """Display name for :attr:`nd_class`."""
        return nd_class_name(self.nd_class)

    @property
    def effect_pattern_name(self) -> str:
        """Display name for :attr:`effect_pattern`."""
        return effect_pattern_name(self.effect_pattern)

    def to_dict(self) -> dict:
        """Field-for-field the CLI ``--json`` shape (the tri-surface parity
        contract — byte-identical to the TS ``MoteDetail.toJSON()``)."""
        return {
            "mote_id": self.mote_id,
            "mote_def_hash": self.mote_def_hash,
            "def_found": self.def_found,
            "step_kind": self.step_kind,
            "model_id": self.model_id,
            "prompt": self.prompt,
            "prompt_truncated": self.prompt_truncated,
            "config_subset": [e.to_dict() for e in self.config_subset],
            "tool_contract": dict(sorted(self.tool_contract.items())),
            "logic_ref": self.logic_ref,
            "nd_class": self.nd_class_name,
            "effect_pattern": self.effect_pattern_name,
            "critic_for": self.critic_for,
            "is_topology_shaper": self.is_topology_shaper,
            "schema_version": self.schema_version,
        }

File: bindings/typescript/src/motes.ts
/**
 * The per-mote definition view — `GetMoteDetail` (Batch B). DISPLAY-ONLY
 * (SN-8): the capped definition summary the coordinator persisted at admission,
 * resolved by `mote_def_hash`; nothing here authorizes anything. A mote that
 * has not committed (or was admitted by a pre-Batch-B binary) answers
 * `defFound: false` honestly. Kept in its own module per the module-per-concern
 * discipline (the `runs.ts` pattern).
 */

import type {
  MoteConfigEntry as PbMoteConfigEntry,
  MoteDetail as PbMoteDetail,
} from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** Display name for a wire `NdClass` discriminant (CLI-parity strings). */
export function ndClassName(nd: number): string {
  switch (nd) {
    case 1:
      return "PURE";
    case 2:
      return "READ_ONLY_NONDET";
    case 3:
      return "WORLD_MUTATING";
    default:
      return "UNKNOWN";
  }
}

/** Display name for a wire `EffectPattern` discriminant (CLI-parity strings). */
export function effectPatternName(ep: number): string {
  switch (ep) {
    case 1:
      return "IdempotentByConstruction";
    case 2:
      return "StageThenCommit";
    case 3:
      return "ValidateThenCommit";
    default:
      return "UNKNOWN";
  }
}

/** One capped config entry of a Mote definition (opaque display bytes). */
export class MoteConfigItem {
  constructor(
    readonly key: string,
    /** The (possibly truncated) raw value bytes. */
    readonly value: Uint8Array,
    readonly truncated: boolean,
    readonly fullLen: number,
  ) {}

  static fromProto(e: PbMoteConfigEntry): MoteConfigItem {
    return new MoteConfigItem(e.key, e.value, e.truncated, Number(e.fullLen));
  }

  /** A plain snake_case object (the CLI `--json` parity shape). */
  toJSON() {
    return {
      key: this.key,
      value_hex: encode(this.value),
      truncated: this.truncated,
      full_len: this.fullLen,
    };
  }
}

/** The capped, display-only definition summary of one Mote. */
export class MoteDetail {
  constructor(
    readonly moteId: string,
    /** Hex def hash; EMPTY string until the Mote commits. */
    readonly moteDefHash: string,
    /** `false`: uncommitted, or admitted by a pre-Batch-B binary. */
    readonly defFound: boolean,
    /** `"pure" | "model" | "exec" | "shaper" | "critic" | "react-turn"` (display). */
    readonly stepKind: string,
    readonly modelId: string,
    /** The instruction text (`config_subset["prompt"]`), capped server-side. */
    readonly prompt: string,
    readonly promptTruncated: boolean,
    readonly configSubset: MoteConfigItem[],
    /** Tool name → pinned version. */
    readonly toolContract: Record<string, string>,
    readonly logicRef: string,
    readonly ndClass: number,
    readonly effectPattern: number,
    /** Hex producer id this Mote critiques, or `undefined`. */
    readonly criticFor: string | undefined,
    readonly isTopologyShaper: boolean,
    readonly schemaVersion: number,
  ) {}

  static fromProto(d: PbMoteDetail): MoteDetail {
    return new MoteDetail(
      encode(d.moteId),
      encode(d.moteDefHash),
      d.defFound,
      d.stepKind,
      d.modelId,
      d.prompt,
      d.promptTruncated,
      d.configSubset.map((e) => MoteConfigItem.fromProto(e)),
      { ...d.toolContract },
      encode(d.logicRef),
      d.ndClass,
      d.effectPattern,
      d.criticFor === undefined ? undefined : encode(d.criticFor),
      d.isTopologyShaper,
      d.schemaVersion,
    );
  }

  /** Display name for {@link ndClass}. */
  get ndClassName(): string {
    return ndClassName(this.ndClass);
  }

  /** Display name for {@link effectPattern}. */
  get effectPatternName(): string {
    return effectPatternName(this.effectPattern);
  }

  /** A plain snake_case object — field-for-field the CLI `--json` shape (the
   *  tri-surface parity contract). */
  toJSON() {
    const tools: Record<string, string> = {};
    for (const key of Object.keys(this.toolContract).sort()) {
      const version = this.toolContract[key];
      if (version !== undefined) {
        tools[key] = version;
      }
    }
    return {
      mote_id: this.moteId,
      mote_def_hash: this.moteDefHash,
      def_found: this.defFound,
      step_kind: this.stepKind,
      model_id: this.modelId,
      prompt: this.prompt,
      prompt_truncated: this.promptTruncated,
      config_subset: this.configSubset.map((e) => e.toJSON()),
      tool_contract: tools,
      logic_ref: this.logicRef,
      nd_class: this.ndClassName,
      effect_pattern: this.effectPatternName,
      critic_for: this.criticFor ?? null,
      is_topology_shaper: this.isTopologyShaper,
      schema_version: this.schemaVersion,
    };
  }
}
